django/test/runner.py

```diff
@@ -776,60 +776,46 @@ def reorder_suite(suite, classes, reverse=False):
-    reordered_suite = suite_class()
-    for tests in bins:
-        reordered_suite.addTests(tests)
-    return reordered_suite
-
-
-def partition_suite_by_type(suite, classes, bins, reverse=False):
-    """
-    Partition a test suite by test type. Also prevent duplicated tests.
-
-    classes is a sequence of types
-    bins is a sequence of TestSuites, one more than classes
-    reverse changes the ordering of tests within bins
+    bins = [OrderedSet() for i in range(len(classes) + 1)]
+    *class_bins, last_bin = bins
 
-    Tests of type classes[i] are added to bins[i],
-    tests with no match found in classes are place in bins[-1]
-    """
     for test in iter_test_cases(suite, reverse=reverse):
-        for i in range(len(classes)):
-            if isinstance(test, classes[i]):
-                bins[i].add(test)
+        for test_bin, test_class in zip(class_bins, classes):
+            if isinstance(test, test_class):
                 break
         else:
-            bins[-1].add(test)
+            test_bin = last_bin
+        test_bin.add(test)
+
+    suite_class = type(suite)
+    return suite_class(itertools.chain(*bins))
 
 
 def partition_suite_by_case(suite):
     """Partition a test suite by test case, preserving the order of tests."""
-    subsuites = []
     suite_class = type(suite)
-    tests = iter_test_cases(suite)
-    for test_type, test_group in itertools.groupby(tests, type):
-        subsuite = suite_class(test_group)
-        subsuites.append(subsuite)
+    all_tests = iter_test_cases(suite)
+    return [
+        suite_class(tests) for _, tests in itertools.groupby(all_tests, type)
+    ]
+
 
-    return subsuites
+def does_test_match_tags(test, tags, exclude_tags):
+    test_tags = set(getattr(test, 'tags', []))
+    test_fn_name = getattr(test, '_testMethodName', str(test))
+    if hasattr(test, test_fn_name):
+        test_fn = getattr(test, test_fn_name)
+        test_fn_tags = list(getattr(test_fn, 'tags', []))
+        test_tags = test_tags.union(test_fn_tags)
+    matched_tags = test_tags.intersection(tags)
+    return (
+        (matched_tags or not tags) and not test_tags.intersection(exclude_tags)
+    )
 
 
 def filter_tests_by_tags(suite, tags, exclude_tags):
     suite_class = type(suite)
-    filtered_suite = suite_class()
-
-    for test in iter_test_cases(suite):
-        test_tags = set(getattr(test, 'tags', set()))
-        test_fn_name = getattr(test, '_testMethodName', str(test))
-        test_fn = getattr(test, test_fn_name, test)
-        test_fn_tags = set(getattr(test_fn, 'tags', set()))
-        all_tags = test_tags.union(test_fn_tags)
-        matched_tags = all_tags.intersection(tags)
-        if (matched_tags or not tags) and not all_tags.intersection(exclude_tags):
-            filtered_suite.addTest(test)
-
-    return filtered_suite
+    return suite_class(
+        test for test in iter_test_cases(suite)
+        if does_test_match_tags(test, tags, exclude_tags)
```

Comment:
I would chop `does_`:
```suggestion
        if test_match_tags(test, tags, exclude_tags)
```